tools/test/test_translate_annotations.py

```diff
@@ -0,0 +1,153 @@
+# translation should reflect that as well as the line numbers
+
+# $ git clone -b 1.0.2 https://github.com/cscorley/whatthepatch.git
+# $ cd whatthepatch/tests/casefiles
+# $ git diff --no-index --unified=0 lao tzu
+lao_tzu_diff = '''
+diff --git a/lao b/tzu
+index 635ef2c..5af88a8 100644
+--- a/lao
++++ b/tzu
+@@ -1,2 +0,0 @@
+-The Way that can be told of is not the eternal Way;
+-The name that can be named is not the eternal name.
+@@ -4 +2,2 @@ The Nameless is the origin of Heaven and Earth;
+-The Named is the mother of all things.
++The named is the mother of all things.
++
+@@ -11,0 +11,3 @@ But after they are produced,
++They both may be called deep and profound.
++Deeper and more profound,
++The door of all subtleties!
+'''.lstrip()
+
+sparser_diff = '''
+diff --git a/foo.txt b/bar.txt
+index 27a6dad..6fae323 100644
+--- a/foo.txt
++++ b/bar.txt
+@@ -4,3 +4,2 @@ lines
+-lines
+-lines
+-lines
++A change!!
++Wow
+@@ -10,2 +8,0 @@ more lines
+-even more
+-even more
+'''.lstrip()
+
+
+class TestTranslateAnnotations(unittest.TestCase):
+    maxDiff = None
+
+    def test_parse_diff(self) -> None:
+        self.assertEqual(
+            parse_diff(lao_tzu_diff),
+            {
+                'old_filename': 'lao',
+                'hunks': [
+                    {
+                        'old_start': 1,
+                        'old_count': 2,
+                        'new_start': 0,
+                        'new_count': 0,
+                    },
+                    {
+                        'old_start': 4,
+                        'old_count': 1,
+                        'new_start': 2,
+                        'new_count': 2,
+                    },
+                    {
+                        'old_start': 11,
+                        'old_count': 0,
+                        'new_start': 11,
+                        'new_count': 3,
+                    },
+                ],
+            },
+        )
+
+    def test_translate_lao_tzu(self) -> None:
+        diff = parse_diff(lao_tzu_diff)
+        self.assertEqual(translate(diff, -1), None)  # out of bounds
+        self.assertEqual(translate(diff, 0), None)  # we start at 1
+        self.assertEqual(translate(diff, 1), 3)
+        self.assertEqual(translate(diff, 2), None)
+        self.assertEqual(translate(diff, 3), None)
+        self.assertEqual(translate(diff, 4), 5)
+        self.assertEqual(translate(diff, 5), 6)
+        self.assertEqual(translate(diff, 6), 7)
+        self.assertEqual(translate(diff, 7), 8)
+        self.assertEqual(translate(diff, 8), 9)
+        self.assertEqual(translate(diff, 9), 10)
+        self.assertEqual(translate(diff, 10), 11)
+        self.assertEqual(translate(diff, 11), None)
+        self.assertEqual(translate(diff, 12), None)
+        self.assertEqual(translate(diff, 13), None)
+        self.assertEqual(translate(diff, 14), 12)  # keep going
+        self.assertEqual(translate(diff, 15), 13)
```

Comment:
Is there some way to group these by category? 

Like have a section for out of bounds, a section where the translate produces None, etc? This test case doesn't quite illuminate what's supposed to happen when you use translate. (Not a big deal though, I'd just read the code)

Reply:
Yeah I think that's a good idea! I'll add some comments to do exactly that

Reply:
Ok, I added a bunch more comments: https://github.com/pytorch/pytorch/pull/55569/commits/71b09fe4787ee33ab3664cc3d9218ceaf37d0d6b does this help clarify things better?